### PR TITLE
Data repositories

### PR DESCRIPTION
**Pull request by C-reshetka**

### CODE REVIEW

File: common/data/dataHelpers.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿/* eslint-disable */
+export const StrapiBaseUrl = "https://map-api.ekaterinburg.io/api";
+
+export async function getDataJsonByUrl(url: string){
+    return await fetch(url)
+        .then(x => x.json());
+}
```
Comment (a-chabin) [inline common/data/dataHelpers.ts:7]: Let's remove this function and use fetch everywhere

File: common/data/dataHelpers.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+﻿/* eslint-disable */
+export const StrapiBaseUrl = "https://map-api.ekaterinburg.io/api";
+
+export async function getDataJsonByUrl(url: string){
+    return await fetch(url)
+        .then(x => x.json());
```
Comment (a-chabin) [inline common/data/dataHelpers.ts:6]: We should to use only `await` or only `then`.

await example:

const response = await fetch(url);
const data = await response.json();

File: common/data/base/houseBase.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+﻿/* eslint-disable */
+import {getDataJsonByUrl, getObjectsTotalCount, StrapiBaseUrl} from "../dataHelpers";
+import {CanGetById} from "./canGetById";
+
+
+export class HouseBase extends CanGetById{
```
Comment (a-chabin) [inline common/data/base/houseBase.ts:6]: Let's simplify classes to objects for better readability

```
export const houseBase = {
    getObject: () => {
       return ...
    },
    getObjectsPolygonsByRange: () => {
       return ...
    }
}
``

File: common/data/base/canGetById.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+﻿/* eslint-disable */
```
Comment (a-chabin) [inline common/data/base/canGetById.ts:1]: Let's fix the ESLint warnings/errors

File: common/data/base/houseBase.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    public override async getObject(id: string): Promise<any> {
```
Comment (a-chabin) [inline common/data/base/houseBase.ts:7]: Let's describe the type `<any>`

File: common/data/designCode/designCode.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        for (const e of this.inputData){
+            if (e.id !== id)
+                continue;
+            this.objectById[id] = e;
+            return e;
+        }
+    }
+    
+    private async getAndSaveData() {
+        await fetch("https://map.ekaterinburg.design/api/map")
```
Comment (a-chabin) [inline common/data/designCode/designCode.ts:67]: https://github.com/ekaterinburgdev/map/pull/7/files#r1084812177

File: common/data/dataHelpers.ts (hunk 1 above)
Comment (C-reshetka) [reply common/data/dataHelpers.ts:7]: but i'll write it everywhere

File: common/data/dataHelpers.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿/* eslint-disable */
+export const StrapiBaseUrl = "https://map-api.ekaterinburg.io/api";
```
Comment (a-chabin) [inline common/data/dataHelpers.ts:2]: nit:

Let's rename this const 

`StrapiBaseUrl` → `STRAPI_BASE_URL`

File: common/data/dtp/dtp.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+
+export class Dtp extends CanGetById{
+    public override async getObject(id: string): Promise<any> {
+        return super.getObject(id, "/dtps");
+    }
+
+    public async getObjectsBySeverity(type: DtpSeverityType){
+        const totalCount = await getObjectsTotalCount("/dtps");
+        return (await getDataJsonByUrl(StrapiBaseUrl +
```
Comment (a-chabin) [inline common/data/dtp/dtp.ts:15]: Let's not mix `${string-concat}` and `+`

`${StrapiBaseUrl}/dtps?filters[severity][$eq]=${type}&populate=data,geometry&pagination[pageSize]=${totalCount}`

File: common/data/dataHelpers.ts (hunk 1 above)
Comment (a-chabin) [reply common/data/dataHelpers.ts:7]: ok, I saw

Let's call it `fetchAPI()`

Comment (a-chabin) [reply common/data/dataHelpers.ts:7]:
```js
async function fetchAPI(url) {
   const response = await fetch(url);
   const data = await response.json();
}
```

File: common/data/designCode/designCode.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    private readonly objectsIdsByType: Map<string, string[]>
+    private readonly objectById: Map<string, DesignCodeObject>;
+    private inputData?: any;
+    
+    constructor() {
+        this.objectsIdsByType = new Map<string, string[]>();
+        this.objectById = new Map<string, DesignCodeObject>();
+    }
+    
+    public async getFilters(): Promise<string[]> {
```
Comment (a-chabin) [inline common/data/designCode/designCode.ts:16]: Is this a `Promise' for sure?

File: common/data/base/houseBase.ts (hunk 3 above)
Comment (C-reshetka) [reply common/data/base/houseBase.ts:6]: maybe if house is a class, then the base should also be one?

File: common/data/designCode/designCode.ts (hunk 9 above)
Comment (a-chabin) [reply common/data/designCode/designCode.ts:16]: I suggest returning no Promise everywhere in these functions

File: common/data/okn/okn.ts [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+import { Area } from '../base/objectsBase';
+import { OknAreaType, OknObjectSignificanceType } from './oknConstants';
+import { OknObject, OknObjectWithGeometry } from './oknObject';
+
+export const okn = {
+    async getObject(id: string): Promise<OknObject> {
+        return canGetById.getObject(id, '/okn-objects')
+            .then((x) => {
+                // eslint-disable-next-line no-param-reassign
+                x.attributes.img = x.attributes?.img.split("','")[0].slice(8);
```
Comment (hiba9201) [inline common/data/okn/okn.ts:13]: Can we just replace single quotes to double quotes, parse json and get image?`JSON.parse(x.attributes.img.replaceAll("'", '"'))`

File: common/data/okn/okn.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    async getAreaByType(type: OknAreaType): Promise<Area | undefined> {
+        const totalCount = await getObjectsTotalCount(`${STRAPI_BASE_URL}/okn-objects`);
+        if (type === OknAreaType.ObjectZone) {
+            return (await fetchAPI(`${STRAPI_BASE_URL
+            }/okn-objects?populate=geometry,data,borders&pagination[pageSize]=${totalCount}`))
+                .data.map((x) => x.attributes?.borders?.coordinates);
+        } if (type === OknAreaType.ProtectZone) {
+            return (await fetchAPI(`${STRAPI_BASE_URL
+            }/okn-protect-zones?populate=geometry,data&pagination[pageSize]=${totalCount}`))
+                .data.map((x) => x.attributes.geometry.coordinates[0]);
```
Comment (hiba9201) [inline common/data/okn/okn.ts:34]: Why we take only first element? What is in those fields?

File: common/data/base/houseBase.ts [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+export const houseBase = {
+    async getObject(id: string): Promise<HouseObject> {
+        return canGetById.getObject(id, '/house');
+    },
+    async getObjectsPolygonsByRange(from: string, to: string, filterName: string) {
+        const totalCount = await getObjectsTotalCount(`${STRAPI_BASE_URL}/house`);
+        return (await fetchAPI(`${STRAPI_BASE_URL
+        }/house?populate=borders&filters[${filterName}][$gte]=${from}&filters[${filterName}][$lte]=${to}&pagination[pageSize]=${totalCount}`))
+            .data.map((x) => x.attributes.borders?.coordinates);
```
Comment (hiba9201) [inline common/data/base/houseBase.ts:12]: I just saw that in strapi those coordinates are srtings, can you cast them to numbers here?

Comment (hiba9201) [reply common/data/base/houseBase.ts:12]: No need! I just asked Olya to fix it in strapi

Comment (hiba9201) [reply common/data/base/houseBase.ts:12]: Also Olya is going to switch places of coordinates, so we don't need to do it here anymore

File: common/data/okn/oknObject.ts [hunk 13]
```diff
@@ -19,7 +19,7 @@ export interface OknAttributes {
     date: string;
     type: string;
     category: OknObjectSignificanceType;
-    img?: string;
+    img?: any;
```
Comment (hiba9201) [inline common/data/okn/oknObject.ts:22]: Images have type { url: string; title: string }. That's what I see in okn objects in strapi

File: common/data/base/objectsBase.ts [hunk 14]
```diff
@@ -10,10 +10,9 @@ export interface ImageParameters {
     src: string;
 }
 
-export interface Point {
-    lat: string,
-    lng: string,
-}
+export type Point = [
+    string, string,
```
Comment (hiba9201) [inline common/data/base/objectsBase.ts]: `[number, number]`?

Comment (C-reshetka) [reply common/data/base/objectsBase.ts]: oh, sry

File: common/data/base/canGetById.ts [hunk 15]
```diff
@@ -0,0 +1,7 @@
+import { fetchAPI, STRAPI_BASE_URL } from '../dataHelpers';
+
+export const canGetById = {
```
Comment (hiba9201) [inline common/data/base/canGetById.ts:3]: Why `can`? Are we checking something here?

File: common/data/okn/okn.ts [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+import { Area } from '../base/objectsBase';
+import { OknAreaType, OknObjectSignificanceType } from './oknConstants';
+import { OknObject, OknObjectWithGeometry } from './oknObject';
+
+export const okn = {
+    async getObject(id: string): Promise<OknObject> {
+        return canGetById.getObject(id, '/okn-objects')
+            .then((x) => {
+                // eslint-disable-next-line no-param-reassign
+                x.attributes.img = JSON.parse(x.attributes?.img.replaceAll("'", '"'));
```
Comment (hiba9201) [inline common/data/okn/okn.ts:13]: I think we should do the same with `document` field. It's the same story: json in string with single quotes like images

File: common/data/okn/oknObject.ts [hunk 17]
```diff
@@ -20,7 +20,7 @@ export interface OknAttributes {
     type: string;
     category: OknObjectSignificanceType;
     img?: { url: string; title: string };
-    document: string;
+    document: { arcive: { id: number; url: string }; date: string; name: string; number: string }[];
```
Comment (hiba9201) [inline common/data/okn/oknObject.ts:23]: archive

File: components/Map/MainContainer/MapMainContainer.tsx [hunk 18]
```diff
@@ -55,6 +56,8 @@ function MapMainContainer({ placemarksData, showFilterHeading = true }: Props) {
             .map((m) => ({ ...m, isOpen: m.id === popup?.id })),
         [placemarks, selectedMarksTypes, popup?.id],
     );
+    
+    okn.getObject("93").then(console.log)
```
Comment (hiba9201) [inline components/Map/MainContainer/MapMainContainer.tsx]: Debug?

File: common/data/okn/oknObject.ts (hunk 17 above)
Comment (C-reshetka) [reply common/data/okn/oknObject.ts:23]: don't understand

File: components/Map/MainContainer/MapMainContainer.tsx (hunk 18 above)
Comment (C-reshetka) [reply components/Map/MainContainer/MapMainContainer.tsx]: ye, alredy fixed

File: common/data/okn/oknObject.ts (hunk 17 above)
Comment (hiba9201) [reply common/data/okn/oknObject.ts:23]: Typo. `archive`, not `arcive`

**Pull request merged**